src/DynamoCoreWpf/Views/Core/DynamoView.xaml.cs

```diff
@@ -186,11 +189,94 @@ public DynamoView(DynamoViewModel dynamoViewModel)
             this.dynamoViewModel.RequestReturnFocusToView += OnRequestReturnFocusToView;
             FocusableGrid.InputBindings.Clear();
         }
 
+        // This method adds a tab item to the right side bar and 
+        // sets the extension window as the tab content.
+        internal TabItem AddTabItem(IViewExtension viewExtension, ContentControl contentControl)
+        {
+            int count = TabItems.Count;
+
+            if (!IsExtensionAddedToRightSideBar(viewExtension))
+            {
+                tabDynamic.DataContext = null;
+
+                // creates a new tab item
+                TabItem tab = new TabItem();
+                tab.Header = viewExtension.GetType().Name;
+                tab.Tag = viewExtension.ToString();
+                tab.HeaderTemplate = tabDynamic.FindResource("TabHeader") as DataTemplate;
+
+                // setting the extension window to the current tab content
+                tab.Content = contentControl.Content;
+
+                //Insert the tab at the end
+                TabItems.Insert(count, tab);
+                TabItems = TabItems;
+
+                tabDynamic.DataContext = TabItems;
+                tabDynamic.SelectedItem = tab;
+
+                return tab;
+            }
+            return null;
+        }
+
+        // This method triggers the close operation on the selected tab. 
+        private void CloseTab(object sender, RoutedEventArgs e)
+        {
+            string tabName = (sender as Button).CommandParameter.ToString();
+            
+            TabItem tab = tabDynamic.SelectedItem as TabItem;
+
+            if (tab != null)
+            {
+                // get the selected tab
+                TabItem selectedTab = tabDynamic.SelectedItem as TabItem;
+
+                // clear tab control binding and bind to the new tab-list. 
+                tabDynamic.DataContext = null;
+                TabItems.Remove(tab);
+                TabItems = TabItems;
+                tabDynamic.DataContext = TabItems;
+
+                // Highlight previously selected tab. if that is removed then Highlight the first tab
+                if (selectedTab == null || selectedTab.Equals(tab))
+                {
+                    if (TabItems.Count > 0) {
+                        selectedTab = TabItems[0];
+                    }
+                }
+                tabDynamic.SelectedItem = selectedTab;
+            }
+        }
+
+        // This event is triggered when the tabitems list is changed and will show/hide the right side bar accordingly.
+        private void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
+        {
+            this.HideOrShowRightSideBar();
+        }
+
+        private Boolean IsExtensionAddedToRightSideBar(IViewExtension viewExtension)
+        {
+            foreach (TabItem tabItem in TabItems)
+            {
+                if (tabItem.Tag.Equals(viewExtension.ToString()))
```

Comment:
what about if we add the fully qualified typeName as the tag? An extension author might override ToString().

Reply:
then we can do typeof(ViewExtension).toString() ?

Reply:
Yes, good point. Changed this to set the tag value with fully qualified type name.